{
  "type": "issue",
  "issue": {
    "id": 2741254080,
    "node_id": "I_kwDOABII586jZDfA",
    "url": "https://api.github.com/repos/bitcoin/bitcoin/issues/31506",
    "repository_url": "https://api.github.com/repos/bitcoin/bitcoin",
    "labels_url": "https://api.github.com/repos/bitcoin/bitcoin/issues/31506/labels%7B/name%7D",
    "comments_url": "https://api.github.com/repos/bitcoin/bitcoin/issues/31506/comments",
    "events_url": "https://api.github.com/repos/bitcoin/bitcoin/issues/31506/events",
    "html_url": "https://github.com/bitcoin/bitcoin/issues/31506",
    "number": 31506,
    "state": "open",
    "state_reason": null,
    "title": "test: different error message fails rpc_signer.py",
    "body": "### Is there an existing issue for this?\n\n- [X] I have searched the existing issues\n\n### Current behaviour\n\nOn my Ubuntu 24.04 system on running rpc_signer.py functional test (after compiling from source) fails the test which tests the case of missing script (on master and 28.x but passes on 27.x !) \r\nI tried running (with only regtest = 1 in bitcoin.conf)\r\n```\r\nbitcoind  -daemon  -signer=fake.py\r\nbitcoin-cli enumeratesigners\r\n```\r\nto get the following:\r\n```\r\nexecve failed: Not a directory (20)\r\n```\r\nwhich is different than what is asserted in the test.\r\nI tried compiling 28.x and master on a Ubuntu 24.04 VM running on Qemu/KVM and the test passes there.\r\n\r\nI followed all the steps for installing the dependencies mentioned in build-unix.md and checked the versions mentioned in dependencies.md . Also tried removing bitcoin and again compiling from source but the test still fails.\n\n### Expected behaviour\n\nThe functional test should pass just like it does on the CI. \n\n### Steps to reproduce\n\nI referred to build-unix.md for compile instructions and installing dependencies.\r\nFor master followed the following commands:\r\n```\r\ncmake -B build\r\ncmake --build build\r\ncmake --install build\r\nbuild/test/functional/rpc_signer.py\r\n```\r\nwhich gives basically the following error:\r\n```\r\nTraceback (most recent call last):\r\n  File \"/home/codingp110/bitcoin/test/functional/test_framework/test_framework.py\", line 135, in main\r\n    self.run_test()\r\n  File \"/home/codingp110/bitcoin/build/test/functional/rpc_signer.py\", line 56, in run_test\r\n    assert_raises_rpc_error(\r\n  File \"/home/codingp110/bitcoin/test/functional/test_framework/util.py\", line 151, in assert_raises_rpc_error\r\n    assert try_rpc(code, message, fun, *args, **kwds), \"No exception raised\"\r\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\r\n  File \"/home/codingp110/bitcoin/test/functional/test_framework/util.py\", line 166, in try_rpc\r\n    raise AssertionError(\r\nAssertionError: Expected substring not found in error message:\r\nsubstring: 'execve failed: No such file or directory'\r\nerror message: 'execve failed: Not a directory (20)'.\r\n```\r\nFor 28.x and 27.x (used git worktree to create separate directories and then git reset hard to overwrite with upstream/28.x and upstream/27.x):\r\n```\r\n./autogen.sh\r\n./configure\r\nmake\r\ntest/functional/rpc_signer.py\r\n```\r\n28.x gives similar error as master but 27.x passes!\n\n### Relevant log output\n\nrelevant log (master):\r\n```\r\n2024-12-16T02:42:47.859000Z TestFramework (ERROR): Assertion failed\r\nTraceback (most recent call last):\r\n  File \"/home/codingp110/bitcoin/test/functional/test_framework/util.py\", line 160, in try_rpc\r\n    fun(*args, **kwds)\r\n  File \"/home/codingp110/bitcoin/test/functional/test_framework/coverage.py\", line 50, in __call__\r\n    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)\r\n                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\r\n  File \"/home/codingp110/bitcoin/test/functional/test_framework/authproxy.py\", line 146, in __call__\r\n    raise JSONRPCException(response['error'], status)\r\ntest_framework.authproxy.JSONRPCException: execve failed: Not a directory (20) (-1)\r\n\r\nDuring handling of the above exception, another exception occurred:\r\n\r\nTraceback (most recent call last):\r\n  File \"/home/codingp110/bitcoin/test/functional/test_framework/test_framework.py\", line 135, in main\r\n    self.run_test()\r\n  File \"/home/codingp110/bitcoin/build/test/functional/rpc_signer.py\", line 56, in run_test\r\n    assert_raises_rpc_error(\r\n  File \"/home/codingp110/bitcoin/test/functional/test_framework/util.py\", line 151, in assert_raises_rpc_error\r\n    assert try_rpc(code, message, fun, *args, **kwds), \"No exception raised\"\r\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\r\n  File \"/home/codingp110/bitcoin/test/functional/test_framework/util.py\", line 166, in try_rpc\r\n    raise AssertionError(\r\nAssertionError: Expected substring not found in error message:\r\nsubstring: 'execve failed: No such file or directory'\r\nerror message: 'execve failed: Not a directory (20)'.\r\n```\r\n\r\nrunning `bitcoind` with the `-signer=fake.py` and then getting strace output on running `enumeratesigners` rpc gave only the following major differences on master and 27.x\r\non master:\r\n```\r\nreadv(6, [{iov_base=\"HTTP/1.1 200 OK\\r\\nContent-Type: application/json\\r\\nDate: Sat, 14 Dec 2024 09:31:05 GMT\\r\\nContent-Length: 93\\r\\nConnection: close\\r\\n\\r\\n{\\\"jsonrpc\\\":\\\"2.0\\\",\\\"error\\\":{\\\"code\\\":-1,\\\"message\\\":\\\"execve failed: Not a directory (20)\\\"},\\\"id\\\":1}\\n\", iov_len=220}], 1) = 220\r\n```\r\non 27.x\r\n```\r\nreadv(6, [{iov_base=\"HTTP/1.1 500 Internal Server Error\\r\\nContent-Type: application/json\\r\\nDate: Sat, 14 Dec 2024 09:32:19 GMT\\r\\nContent-Length: 96\\r\\nConnection: close\\r\\n\\r\\n{\\\"result\\\":null,\\\"error\\\":{\\\"code\\\":-1,\\\"message\\\":\\\"execve failed: No such file or directory\\\"},\\\"id\\\":1}\\n\", iov_len=242}], 1) = 242\r\n```\r\nbut on doing the same for master and 28.x on a VM(Ubuntu 24.04.1 LTS on VirtualBox) gave the same strace output as my system but with the correct error message.\r\n\r\nPS: The logs/outputs mentioned for master(except for the first one in this section) are old and not for master-b042c4f0538c\n\n### How did you obtain Bitcoin Core\n\nCompiled from source\n\n### What version of Bitcoin Core are you using?\n\nv28.99.0-b042c4f0538c \n\n### Operating system and version\n\nUbuntu 24.04.1 LTS\n\n### Machine specifications\n\nOS: Ubuntu 24.04\r\nArchitecture:             x86_64\r\nCPU(s):                   8\r\nVendor ID:                GenuineIntel\r\n  Model name:             12th Gen Intel(R) Core(TM) i3-1215U",
    "user": {
      "login": "110CodingP",
      "id": 119165932,
      "node_id": "U_kgDOBxpT7A",
      "avatar_url": "https://avatars.githubusercontent.com/u/119165932?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/110CodingP",
      "html_url": "https://github.com/110CodingP",
      "followers_url": "https://api.github.com/users/110CodingP/followers",
      "following_url": "https://api.github.com/users/110CodingP/following%7B/other_user%7D",
      "gists_url": "https://api.github.com/users/110CodingP/gists%7B/gist_id%7D",
      "starred_url": "https://api.github.com/users/110CodingP/starred%7B/owner%7D%7B/repo%7D",
      "subscriptions_url": "https://api.github.com/users/110CodingP/subscriptions",
      "organizations_url": "https://api.github.com/users/110CodingP/orgs",
      "repos_url": "https://api.github.com/users/110CodingP/repos",
      "events_url": "https://api.github.com/users/110CodingP/events%7B/privacy%7D",
      "received_events_url": "https://api.github.com/users/110CodingP/received_events",
      "type": "User",
      "site_admin": false,
      "patch_url": null
    },
    "labels": [
      {
        "id": 62963516,
        "node_id": "MDU6TGFiZWw2Mjk2MzUxNg==",
        "url": "https://api.github.com/repos/bitcoin/bitcoin/labels/Tests",
        "name": "Tests",
        "color": "d4c5f9",
        "default": false
      }
    ],
    "assignees": [],
    "author_association": "NONE",
    "locked": false,
    "comments": 3,
    "created_at": "2024-12-16T03:48:07Z",
    "updated_at": "2024-12-16T11:46:02Z"
  },
  "events": [
    {
      "event": "commented",
      "id": 2544522461,
      "node_id": "IC_kwDOABII586XqlTd",
      "url": "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/2544522461",
      "actor": {
        "login": "110CodingP",
        "id": 119165932,
        "node_id": "U_kgDOBxpT7A",
        "avatar_url": "https://avatars.githubusercontent.com/u/119165932?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/110CodingP",
        "html_url": "https://github.com/110CodingP",
        "followers_url": "https://api.github.com/users/110CodingP/followers",
        "following_url": "https://api.github.com/users/110CodingP/following%7B/other_user%7D",
        "gists_url": "https://api.github.com/users/110CodingP/gists%7B/gist_id%7D",
        "starred_url": "https://api.github.com/users/110CodingP/starred%7B/owner%7D%7B/repo%7D",
        "subscriptions_url": "https://api.github.com/users/110CodingP/subscriptions",
        "organizations_url": "https://api.github.com/users/110CodingP/orgs",
        "repos_url": "https://api.github.com/users/110CodingP/repos",
        "events_url": "https://api.github.com/users/110CodingP/events%7B/privacy%7D",
        "received_events_url": "https://api.github.com/users/110CodingP/received_events",
        "type": "User",
        "site_admin": false,
        "patch_url": null
      },
      "commit_id": null,
      "commit_url": null,
      "created_at": "2024-12-16T03:50:18Z",
      "updated_at": "2024-12-16T03:50:18Z",
      "author_association": "NONE",
      "body": "I understand that this may be a problem with only my system and hence not relevant, happy to close the issue in that case!",
      "user": {
        "login": "110CodingP",
        "id": 119165932,
        "node_id": "U_kgDOBxpT7A",
        "avatar_url": "https://avatars.githubusercontent.com/u/119165932?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/110CodingP",
        "html_url": "https://github.com/110CodingP",
        "followers_url": "https://api.github.com/users/110CodingP/followers",
        "following_url": "https://api.github.com/users/110CodingP/following%7B/other_user%7D",
        "gists_url": "https://api.github.com/users/110CodingP/gists%7B/gist_id%7D",
        "starred_url": "https://api.github.com/users/110CodingP/starred%7B/owner%7D%7B/repo%7D",
        "subscriptions_url": "https://api.github.com/users/110CodingP/subscriptions",
        "organizations_url": "https://api.github.com/users/110CodingP/orgs",
        "repos_url": "https://api.github.com/users/110CodingP/repos",
        "events_url": "https://api.github.com/users/110CodingP/events%7B/privacy%7D",
        "received_events_url": "https://api.github.com/users/110CodingP/received_events",
        "type": "User",
        "site_admin": false,
        "patch_url": null
      },
      "html_url": "https://github.com/bitcoin/bitcoin/issues/31506#issuecomment-2544522461",
      "issue_url": "https://api.github.com/repos/bitcoin/bitcoin/issues/31506"
    },
    {
      "event": "commented",
      "id": 2544847899,
      "node_id": "IC_kwDOABII586Xr0wb",
      "url": "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/2544847899",
      "actor": {
        "login": "maflcko",
        "id": 6399679,
        "node_id": "MDQ6VXNlcjYzOTk2Nzk=",
        "avatar_url": "https://avatars.githubusercontent.com/u/6399679?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/maflcko",
        "html_url": "https://github.com/maflcko",
        "followers_url": "https://api.github.com/users/maflcko/followers",
        "following_url": "https://api.github.com/users/maflcko/following%7B/other_user%7D",
        "gists_url": "https://api.github.com/users/maflcko/gists%7B/gist_id%7D",
        "starred_url": "https://api.github.com/users/maflcko/starred%7B/owner%7D%7B/repo%7D",
        "subscriptions_url": "https://api.github.com/users/maflcko/subscriptions",
        "organizations_url": "https://api.github.com/users/maflcko/orgs",
        "repos_url": "https://api.github.com/users/maflcko/repos",
        "events_url": "https://api.github.com/users/maflcko/events%7B/privacy%7D",
        "received_events_url": "https://api.github.com/users/maflcko/received_events",
        "type": "User",
        "site_admin": false,
        "patch_url": null
      },
      "commit_id": null,
      "commit_url": null,
      "created_at": "2024-12-16T07:57:30Z",
      "updated_at": "2024-12-16T07:57:30Z",
      "author_association": "MEMBER",
      "body": "> 28.x gives similar error as master but 27.x passes!\r\n\r\nIn this case, a git bisect would help to see for which commit the test first fails.",
      "user": {
        "login": "maflcko",
        "id": 6399679,
        "node_id": "MDQ6VXNlcjYzOTk2Nzk=",
        "avatar_url": "https://avatars.githubusercontent.com/u/6399679?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/maflcko",
        "html_url": "https://github.com/maflcko",
        "followers_url": "https://api.github.com/users/maflcko/followers",
        "following_url": "https://api.github.com/users/maflcko/following%7B/other_user%7D",
        "gists_url": "https://api.github.com/users/maflcko/gists%7B/gist_id%7D",
        "starred_url": "https://api.github.com/users/maflcko/starred%7B/owner%7D%7B/repo%7D",
        "subscriptions_url": "https://api.github.com/users/maflcko/subscriptions",
        "organizations_url": "https://api.github.com/users/maflcko/orgs",
        "repos_url": "https://api.github.com/users/maflcko/repos",
        "events_url": "https://api.github.com/users/maflcko/events%7B/privacy%7D",
        "received_events_url": "https://api.github.com/users/maflcko/received_events",
        "type": "User",
        "site_admin": false,
        "patch_url": null
      },
      "html_url": "https://github.com/bitcoin/bitcoin/issues/31506#issuecomment-2544847899",
      "issue_url": "https://api.github.com/repos/bitcoin/bitcoin/issues/31506"
    },
    {
      "event": "labeled",
      "id": 15666033803,
      "node_id": "LE_lADOABII586jZDfAzwAAAAOlxLSL",
      "url": "https://api.github.com/repos/bitcoin/bitcoin/issues/events/15666033803",
      "actor": {
        "login": "maflcko",
        "id": 6399679,
        "node_id": "MDQ6VXNlcjYzOTk2Nzk=",
        "avatar_url": "https://avatars.githubusercontent.com/u/6399679?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/maflcko",
        "html_url": "https://github.com/maflcko",
        "followers_url": "https://api.github.com/users/maflcko/followers",
        "following_url": "https://api.github.com/users/maflcko/following%7B/other_user%7D",
        "gists_url": "https://api.github.com/users/maflcko/gists%7B/gist_id%7D",
        "starred_url": "https://api.github.com/users/maflcko/starred%7B/owner%7D%7B/repo%7D",
        "subscriptions_url": "https://api.github.com/users/maflcko/subscriptions",
        "organizations_url": "https://api.github.com/users/maflcko/orgs",
        "repos_url": "https://api.github.com/users/maflcko/repos",
        "events_url": "https://api.github.com/users/maflcko/events%7B/privacy%7D",
        "received_events_url": "https://api.github.com/users/maflcko/received_events",
        "type": "User",
        "site_admin": false,
        "patch_url": null
      },
      "commit_id": null,
      "commit_url": null,
      "created_at": "2024-12-16T07:57:35Z",
      "label": {
        "name": "Tests",
        "color": "d4c5f9"
      }
    },
    {
      "event": "commented",
      "id": 2545404062,
      "node_id": "IC_kwDOABII586Xt8ie",
      "url": "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/2545404062",
      "actor": {
        "login": "Sjors",
        "id": 10217,
        "node_id": "MDQ6VXNlcjEwMjE3",
        "avatar_url": "https://avatars.githubusercontent.com/u/10217?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/Sjors",
        "html_url": "https://github.com/Sjors",
        "followers_url": "https://api.github.com/users/Sjors/followers",
        "following_url": "https://api.github.com/users/Sjors/following%7B/other_user%7D",
        "gists_url": "https://api.github.com/users/Sjors/gists%7B/gist_id%7D",
        "starred_url": "https://api.github.com/users/Sjors/starred%7B/owner%7D%7B/repo%7D",
        "subscriptions_url": "https://api.github.com/users/Sjors/subscriptions",
        "organizations_url": "https://api.github.com/users/Sjors/orgs",
        "repos_url": "https://api.github.com/users/Sjors/repos",
        "events_url": "https://api.github.com/users/Sjors/events%7B/privacy%7D",
        "received_events_url": "https://api.github.com/users/Sjors/received_events",
        "type": "User",
        "site_admin": false,
        "patch_url": null
      },
      "commit_id": null,
      "commit_url": null,
      "created_at": "2024-12-16T11:46:01Z",
      "updated_at": "2024-12-16T11:46:01Z",
      "author_association": "MEMBER",
      "body": "The error \"execve failed: Not a directory (20)\" is quite strange, since we're not expecting `-signer` to point to a directory.\r\n\r\nMost of the action happens here: https://github.com/bitcoin/bitcoin/blame/master/src/common/run_command.cpp\r\n\r\nAnd the biggest change between v27 and v28 is https://github.com/bitcoin/bitcoin/pull/28981\r\n\r\nI assume you've tried deleting the `build` directory to see if you still get the test error?\r\n\r\nDoes `enumeratesigners` work with the real HWI?",
      "user": {
        "login": "Sjors",
        "id": 10217,
        "node_id": "MDQ6VXNlcjEwMjE3",
        "avatar_url": "https://avatars.githubusercontent.com/u/10217?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/Sjors",
        "html_url": "https://github.com/Sjors",
        "followers_url": "https://api.github.com/users/Sjors/followers",
        "following_url": "https://api.github.com/users/Sjors/following%7B/other_user%7D",
        "gists_url": "https://api.github.com/users/Sjors/gists%7B/gist_id%7D",
        "starred_url": "https://api.github.com/users/Sjors/starred%7B/owner%7D%7B/repo%7D",
        "subscriptions_url": "https://api.github.com/users/Sjors/subscriptions",
        "organizations_url": "https://api.github.com/users/Sjors/orgs",
        "repos_url": "https://api.github.com/users/Sjors/repos",
        "events_url": "https://api.github.com/users/Sjors/events%7B/privacy%7D",
        "received_events_url": "https://api.github.com/users/Sjors/received_events",
        "type": "User",
        "site_admin": false,
        "patch_url": null
      },
      "html_url": "https://github.com/bitcoin/bitcoin/issues/31506#issuecomment-2545404062",
      "issue_url": "https://api.github.com/repos/bitcoin/bitcoin/issues/31506"
    }
  ]
}